File: JavaScript/1498-Number of Subsequences That Satisfy the Given Sum Condition.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var numSubseq = function(nums, target) {
    
    const kMod = 1000000007;
    const n = nums.length;
    let j = n-1, res = 0;
    
    nums.sort(function(a,b){return a-b});
    
    let calPow = new Array(n);
    calPow[0] = 1
    for (let i=1; i<n; i++){
        calPow[i] = calPow[i-1]*2%kMod;
    }
    
    
    for (let i=0; i<n; i++){
        if (nums[i]*2 > target) break;
        while (nums[i] + nums[j] > target){
            j--;
        }
        res += calPow[j-i];
    }
    return res%kMod;
};

File: C++/1498-Number of Subsequences That Satisfy the Given Sum Condition.cpp
using LL = long long;
class Solution {
public:
    int kMod = 1e9+7;
    LL powWithMod(LL a, LL b, int kMod) {
        a %= kMod;
        LL res = 1;

        while(b) {
            if (b&1) res=(res*a)%kMod;
            a=(a*a)%kMod;
            b>>=1;
        }

        return res;
    }
    int numSubseq(vector<int>& nums, int target) {
        sort(begin(nums), end(nums));
        int N = nums.size(), right = N-1;
        long long res = 0;

        for (int i=0; i<N; i++) {
            while (right > i && nums[right]+nums[i] > target) {
                right--;
            }
            if (nums[i] + nums[right] > target) break;
            int count = right-i+1;
            res = (res + powWithMod(2, count-1, kMod))%kMod;
        }

        return res;
    }
};